File: leetcode/medium/63-unique-paths-2/rec-dp.cpp
/**
 * @file rec-dp.cpp
 * @author Vortexx2
 * @brief Problem 63 - Unique Paths II
 * @date 31-07-2021
 * 
 * Runtime - 4 ms
 * Memory Usage - 14.6 MB
 */
#include <iostream>
#include <vector>
#include <deque>
#include <algorithm>
#include <list>
#include <unordered_map>
#include <map>
#include <set>
#include <stack>
#include <queue>
#include <string>

using namespace std;

class Solution
{
public:
  int uniquePathsWithObstacles(vector<vector<int>> &obstacleGrid)
  {
    if (obstacleGrid.empty())
      return 0;

    if (obstacleGrid[obstacleGrid.size() - 1][obstacleGrid[0].size() - 1])
      return 0;

    unordered_map<string, int> dict;

    return helper(obstacleGrid, dict);
  }

  int helper(vector<vector<int>> &obs, unordered_map<string, int> &dict, int i = 0, int j = 0)
  {
    if (obs[i][j])
      return 0;

    if (i == obs.size() - 1 && j == obs[0].size() - 1)
      return 1;

    string key = to_string(i) + "," + to_string(j);
    auto itr = dict.find(key);

    if (itr != dict.end())
      return itr->second;

    if (i == obs.size() - 1)
      dict[key] = helper(obs, dict, i, j + 1);

    else if (j == obs[0].size() - 1)
      dict[key] = helper(obs, dict, i + 1, j);

    else
      dict[key] = helper(obs, dict, i + 1, j) + helper(obs, dict, i, j + 1);
    return dict[key];
  }
};

File: leetcode/medium/63-unique-paths-2/rec-dp.py
"""
  @author Vortexx2
  Problem 63 - Unique Paths II

  Runtime - 48 ms
  Memory Usage - 15.5 MB
"""
from typing import List


class Solution:
  def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:

    if not obstacleGrid:
      return 0

    if obstacleGrid[len(obstacleGrid) - 1][len(obstacleGrid[0]) - 1]:
      return 0

    dic = {}
    return self.helper(obstacleGrid, dic)

  def helper(self, obs: List[List[int]], dic, i: int = 0, j: int = 0):

    if obs[i][j]:
      return 0

    if i == len(obs) - 1 and j == len(obs[0]) - 1:
      return 1

    key = str(i) + "," + str(j)

    if key in dic:
      return dic[key]

    if i == len(obs) - 1:
      dic[key] = self.helper(obs, dic, i, j + 1)

    elif j == len(obs[0]) - 1:
      dic[key] = self.helper(obs, dic, i + 1, j)

    else:
      dic[key] = self.helper(obs, dic, i + 1, j) + \
          self.helper(obs, dic, i, j + 1)

    return dic[key]
